### PR TITLE
Add back caching

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Adds client specific caching back into the codebase. Specifically caches `get_crosslink_committees` and `get_epoch_start_shard`.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java [hunk 1]
```diff
@@ -96,41 +95,41 @@ public static BeaconStateWithCache deepCopy(BeaconStateWithCache state) {
...
-  public void setPreviousEpochCommitteeCount(UnsignedLong count) {
-    this.previousEpochCommitteeCount = count;
+  public void setEpochStartShard(UnsignedLong epoch, UnsignedLong shard){
+    this.epochStartShards.put(epoch, shard);
   }
 
   public void invalidateCache() {
-    this.previousEpochCommitteeCount = UnsignedLong.MAX_VALUE;
-    this.previousTotalBalance = UnsignedLong.MAX_VALUE;
+    // TODO: clean this cache after finalization
```
Comment (dangerousfood) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java:131]: Can we write an issue for this?

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/CrosslinkCommitteeUtil.java [hunk 2]
```diff
@@ -166,25 +179,35 @@ public static Integer get_shuffled_index(int index, int index_count, Bytes32 see
...
-            .mod(UnsignedLong.valueOf(SHARD_COUNT));
-
-    while (check_epoch.compareTo(epoch) > 0) {
-      check_epoch = check_epoch.minus(UnsignedLong.ONE);
-      shard =
-          shard
-              .plus(UnsignedLong.valueOf(SHARD_COUNT))
-              .minus(get_shard_delta(state, check_epoch))
-              .mod(UnsignedLong.valueOf(SHARD_COUNT));
+    if (state instanceof BeaconStateWithCache
```
Comment (dangerousfood) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/CrosslinkCommitteeUtil.java:182]: this is a thing of beauty

**Review by dangerousfood — APPROVED** [review]
lgtm

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java (hunk 1 above)
Comment (cemozerr) [reply ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/BeaconStateWithCache.java:131]: Yep I'll make an issue

**Pull request merged**